File: documentation-examples-kotlin/src/kotlin/dev/typr/foundationskt/docs/core/ParameterizedQuery.kt
package dev.typr.foundationskt.docs.core

import dev.typr.foundationskt.*
import dev.typr.foundationskt.data.*

@Suppress("unused")
object ParameterizedQuery {
    data class City(val name: String, val population: Int)

    val cityCodec: RowCodecNamed<City> =
        RowCodec.namedBuilder<City>()
            .field("name", DuckDbTypes.varchar, City::name)
            .field("population", DuckDbTypes.integer, City::population)
            .build(::City)

    //start
    fun findCityByName(name: String): Operation<City?> =
        sql { "SELECT ${cityCodec.columnList} FROM city WHERE name = ${DuckDbTypes.varchar(name)}" }
            .query(cityCodec.maxOne())
    //stop
}

File: documentation-examples-scala/src/scala/dev/typr/foundationssc/docs/core/ParameterizedQuery.scala
package dev.typr.foundationssc.docs.core
import dev.typr.foundationssc.*
import dev.typr.foundationssc.Fragment.*
import dev.typr.foundationssc.data.*

@SuppressWarnings(Array("unused"))
object ParameterizedQuery:
  case class City(name: String, population: Int)

  val cityCodec: RowCodecNamed[City] = RowCodec
    .namedBuilder[City]()
    .field("name", DuckDbTypes.varchar)(_.name)
    .field("population", DuckDbTypes.integer)(_.population)
    .build(City.apply)

  // start
  def findCityByName(name: String): OperationRead[Option[City]] =
    sql"SELECT ${cityCodec.columnList} FROM city WHERE name = ${DuckDbTypes.text(name)}"
      .query(cityCodec.maxOne())
  // stop
